program/src/state.rs

```diff
@@ -157,6 +162,11 @@ impl ExchangeRate {
         // The result is in Lamports, because the type system considers Rational
         // dimensionless, but in this case `rate` has dimensions stSOL/SOL, so
         // we need to re-wrap the result in the right type.
+        //
+        // LOW
+        // This is a user error, but immediately after earning any rewards, 1 lamport
+        // would result in 0 stSOL, and the program will happily take the user's
+        // lamport without giving anything in return.
```

Comment:
Let’s look at when this can return 0. It returns a product of two rationals (if we upgrade the left-hand-side to a rational), which is zero if at least one of the factors is zero:

* If `amount` is zero then you should get `0` stSOL out. This is working as intended.
* The other case is where `rate` is 0, and the only way for that to happen is when the numerator `st_sol_supply` is 0. But we checked above, and in case `st_sol_supply == StLamports(0)`, we return the input amount nominally converted to stSOL. So here the user does get a nonzero amount of stSOL (unless `amount` was 0, but then it’s again working as intended).

Reply:
The implementation for `Mul` between a token type and `Rational` is:
```
            fn mul(self, other: Rational) -> Result<$TokenLamports> {
                // This multiplication cannot overflow, because we expand the
                // u64s into u128, and u64::MAX * u64::MAX < u128::MAX.
                let result_u128 = ((self.0 as u128) * (other.numerator as u128))
                    .checked_div(other.denominator as u128)
                    .ok_or(ArithmeticError)?;
                u64::try_from(result_u128)
                    .map($TokenLamports)
                    .map_err(|_| ArithmeticError)
            }
```
If you have `sol_balance = 101` and `st_sol_supply = 100`, then for 1 lamport you'll have `1 * 100 / 101`, which gives 0.  Is that correct?

Either way, it's not a big deal, since you'll spend way more on transaction fees by depositing 1 lamport.

Reply:
Ah, this is a loss due to rounding errors, that can happen indeed. If `sol_balance` ≈ `st_sol_supply` then it is not a problem. But if `st_sol_supply` is much larger than `sol_balance`, then it can become an issue, because in a sense it causes us to lose precision, and users will lose some funds to rounding errors. This is a known problem, but it can be prevented by ensuring that we deposit enough SOL initially at the 1:1 exchange rate.